File: src/org/genyris/interp/Environment.java
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.interp;

import org.genyris.core.DynamicSymbol;
import org.genyris.core.Exp;
import org.genyris.core.Internable;
import org.genyris.core.SimpleSymbol;
import org.genyris.core.Symbol;
import org.genyris.exception.GenyrisException;

public interface Environment {

    Exp lookupVariableValue(Symbol symbol) throws UnboundException;
    Exp lookupDynamicVariableValue(DynamicSymbol symbol) throws UnboundException;
    Exp lookupLexicalVariableValue(SimpleSymbol symbol) throws UnboundException;

    void defineVariable(Symbol symbol, Exp valu) throws GenyrisException;
    void defineLexicalVariable(SimpleSymbol symbol, Exp valu) throws GenyrisException;
    void defineDynamicVariable(DynamicSymbol symbol, Exp valu) throws GenyrisException;

    void setVariableValue(Symbol symbol, Exp valu) throws UnboundException;
    void setLexicalVariableValue(SimpleSymbol symbol, Exp valu) throws UnboundException;
    void setDynamicVariableValue(DynamicSymbol symbol, Exp valu) throws UnboundException;


    String toString();

    Exp lookupInThisClassAndSuperClasses(DynamicSymbol symbol) throws UnboundException;

    SimpleSymbol getNil();

    Symbol internString(String symbolName);

    Exp getSelf() throws UnboundException;

	Internable getSymbolTable();
    boolean isBound(Symbol s);


}

File: src/org/genyris/interp/Environment.kt
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.interp

import org.genyris.core.*
import org.genyris.exception.GenyrisException

interface Environment {
    @kotlin.Throws(UnboundException::class)
    fun lookupVariableValue(symbol: Symbol?): Exp?

    @kotlin.Throws(UnboundException::class)
    fun lookupDynamicVariableValue(symbol: DynamicSymbol?): Exp?

    @kotlin.Throws(UnboundException::class)
    fun lookupLexicalVariableValue(symbol: SimpleSymbol?): Exp?

    @kotlin.Throws(GenyrisException::class)
    fun defineVariable(symbol: Symbol?, valu: Exp?)

    @kotlin.Throws(GenyrisException::class)
    fun defineLexicalVariable(symbol: SimpleSymbol?, valu: Exp?)

    @kotlin.Throws(GenyrisException::class)
    fun defineDynamicVariable(symbol: DynamicSymbol?, valu: Exp?)

    @kotlin.Throws(UnboundException::class)
    fun setVariableValue(symbol: Symbol?, valu: Exp?)

    @kotlin.Throws(UnboundException::class)
    fun setLexicalVariableValue(symbol: SimpleSymbol?, valu: Exp?)

    @kotlin.Throws(UnboundException::class)
    fun setDynamicVariableValue(symbol: DynamicSymbol?, valu: Exp?)


    override fun toString(): String

    @kotlin.Throws(UnboundException::class)
    fun lookupInThisClassAndSuperClasses(symbol: DynamicSymbol?): Exp?

    val nil: SimpleSymbol?

    fun internString(symbolName: String?): Symbol?

    @get:kotlin.Throws(UnboundException::class)
    val self: Exp?

    val symbolTable: Internable?
    fun isBound(s: Symbol?): Boolean
}